UserCode/CodeGPU/plugins/HeterogeneousHGCalHEFRecHitProducer.cc

```diff
@@ -0,0 +1,156 @@
+  vdata_.cce                = ps.getParameter<edm::ParameterSet>("HGCHEF_cce").getParameter<std::vector<double> >("values");
+  vdata_.noise_fC           = ps.getParameter<edm::ParameterSet>("HGCHEF_noise_fC").getParameter<std::vector<double> >("values");
+  vdata_.rcorr              = ps.getParameter< std::vector<double> >("rcorr");
+  vdata_.weights            = ps.getParameter< std::vector<double> >("weights");
+  cdata_.fhOffset_          = ps.getParameter<uint32_t>("offset"); //ddd_->layers(true);
+  cdata_.s_hgcHEF_fCPerMIP_ = vdata_.fCPerMIP.size();
+  cdata_.s_hgcHEF_cce_      = vdata_.cce.size();
+  cdata_.s_hgcHEF_noise_fC_ = vdata_.noise_fC.size();
+  cdata_.s_rcorr_           = vdata_.rcorr.size();
+  cdata_.s_weights_ = vdata_.weights.size();
+  cdata_.hgchefUncalib2GeV_ = 1e-6 / cdata_.hgcHEF_keV2DIGI_;
+  vdata_.waferTypeL = {0, 1, 2};//ddd_->retWaferTypeL(); if depends on geometry the allocation is tricky!
+  cdata_.s_waferTypeL_ = vdata_.waferTypeL.size();
+
+  begin = std::chrono::steady_clock::now();
+
+  tools_.reset(new hgcal::RecHitTools());
+  stride_ = ( (nhitsmax_-1)/32 + 1 ) * 32; //align to warp boundary
+
+  allocate_memory_();
+
+  convert_constant_data_(h_kcdata_);
+
+  produces<HGChefRecHitCollection>(collection_name_);
+}
+
+HeterogeneousHGCalHEFRecHitProducer::~HeterogeneousHGCalHEFRecHitProducer()
+{
+  delete kmdata_;
+  delete h_kcdata_;
+  delete d_kcdata_;
+  delete old_soa_;
+  delete d_oldhits_;
+  delete d_newhits_;
+  delete d_newhits_final_;
+  delete h_newhits_;
+
+  end = std::chrono::steady_clock::now();
+  std::cout << "Time difference (heterogeneous) = " << std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count() << " [ms]" << std::endl;
+}
+
+void HeterogeneousHGCalHEFRecHitProducer::acquire(edm::Event const& event, edm::EventSetup const& setup, edm::WaitingTaskWithArenaHolder w) {
+  const cms::cuda::ScopedContextAcquire ctx{event.streamID(), std::move(w), ctxState_};
+  set_geometry_(setup);
+  event.getByToken(token_, handle_hef_);
+  const auto &hits_hef = *handle_hef_;
+
+  unsigned int nhits = hits_hef.size();
+  convert_collection_data_to_soa_(hits_hef, old_soa_, nhits);
+
+  kmdata_ = new KernelModifiableData<HGCUncalibratedRecHitSoA, HGCRecHitSoA>(nhitsmax_, stride_, old_soa_, d_oldhits_, d_newhits_, d_newhits_final_, h_newhits_);
+  KernelManagerHGCalRecHit kernel_manager(kmdata_);
+  kernel_manager.run_kernels(h_kcdata_, d_kcdata_);
+  new_soa_ = kernel_manager.get_output();
+
+  rechits_ = std::make_unique<HGCRecHitCollection>();
+  convert_soa_data_to_collection_(*rechits_, new_soa_, nhits);
+}
+
+void HeterogeneousHGCalHEFRecHitProducer::produce(edm::Event& event, const edm::EventSetup& setup)
+{
+  cms::cuda::ScopedContextProduce ctx{ctxState_}; //only for GPU to GPU producers
+  event.put(std::move(rechits_), collection_name_);
+}
+
+void HeterogeneousHGCalHEFRecHitProducer::allocate_memory_()
```

Comment:
remove trailing underscore?

Reply:
The underscore shows that the function is a private method of the class. As far as I could read, it does not go against the CMS style code guide, and it is consistent with the nomenclature in the same file, and hopefully in the rest of the code.